File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/ImportApiAnalysis.scala
package org.enso.compiler.pass.analyse

import org.enso.interpreter.runtime.Module
import org.enso.compiler.context.{InlineContext, ModuleContext}
import org.enso.compiler.core.IR
import org.enso.compiler.data.BindingsMap
import org.enso.compiler.pass.IRPass
import org.enso.compiler.pass.desugar.{
  ComplexType,
  FunctionBinding,
  GenerateMethodBodies
}
import org.enso.compiler.pass.resolve.{MethodDefinitions, Patterns}

import scala.annotation.unused

/** Recognizes all defined bindings in the current module and constructs
  * a mapping data structure that can later be used for symbol resolution.
  */
case object ImportApiAnalysis extends IRPass {

  override type Metadata = BindingsMap

  /** The type of configuration for the pass. */
  override type Config = IRPass.Configuration.Default

  /** The passes that this pass depends _directly_ on to run. */
  override val precursorPasses: Seq[IRPass] =
    Seq(ComplexType, FunctionBinding, GenerateMethodBodies)

  /** The passes that are invalidated by running this pass. */
  override val invalidatedPasses: Seq[IRPass] =
    Seq(MethodDefinitions, Patterns)

  /** Executes the pass on the provided `ir`, and returns a possibly transformed
    * or annotated version of `ir`.
    *
    * @param ir            the Enso IR to process
    * @param moduleContext a context object that contains the information needed
    *                      to process a module
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runModule(
    ir: IR.Module,
    moduleContext: ModuleContext
  ): IR.Module = {
    val map: BindingsMap =
      ir.unsafeGetMetadata(BindingAnalysis, "Should exist.")
    val forbiddenImports = new java.util.HashMap[String, String]

    map.resolvedImports.map { imp: BindingsMap.ResolvedImport =>
      val maybeModule = imp.target match {
        case rm: BindingsMap.ResolvedModule =>
          None
          rm.module match {
            case c: BindingsMap.ModuleReference.Concrete => Some(c.module)
            case _                                       => None
          }
        case _ => None
      }

      maybeModule.map(m => {
        val pkg = m.getPackage()
        if (moduleContext.module.getPackage() != pkg) {
          if (pkg.mainFile != null) {
            val found = pkg.findModule("Main")
            if (found != null) {
              // if the package has Main file, then imports has to go thru the main module
              val mainModule = found.asInstanceOf[Module]
              if (m != mainModule) {
                // if different that Main module is requested, do a check
                if (mainModule.getIr == null) {
                  // if main module IR isn't loaded, then certainly the import didn't go thru Main
                  forbiddenImports.put(
                    m.getName().toString(),
                    mainModule.getName().toString()
                  )
                } else {
                  val mainMap: BindingsMap = mainModule.getIr
                    .unsafeGetMetadata(BindingAnalysis, "Should exist.")
                  val checks = for {
                    // go thru all re-exported modules in main
                    (name, modules) <- mainMap.exportedSymbols
                    module          <- modules
                  } yield {
                    module match {
                      case BindingsMap.ResolvedModule(
                            BindingsMap.ModuleReference.Concrete(allowed)
                          ) => {
                        // check if one of the exported modules matches the imported one
                        allowed == m
                      }
                      case _ => false
                    }
                  }
                  if (!checks.exists(b => b)) {
                    // if no re-exported module patches m, report an error
                    forbiddenImports.put(
                      m.getName().toString(),
                      mainModule.getName().toString()
                    )
                  }
                }
              }
            }
          }
        }
      })
    }
    val replace = ir.imports.map(imp =>
      imp match {
        case m: IR.Module.Scope.Import.Module =>
          val name: String = m.name.name
          val pkg          = forbiddenImports.get(name)
          if (pkg != null) {
            IR.Error.ImportExport(
              imp,
              IR.Error.ImportExport
                .SymbolDoesNotExist(name.split("\\.").last, pkg)
            )
          } else {
            m
          }
        case i => i
      }
    )
    ir.copy(imports = replace)
  }

  /** Executes the pass on the provided `ir`, and returns a possibly transformed
    * or annotated version of `ir` in an inline context.
    *
    * @param ir            the Enso IR to process
    * @param inlineContext a context object that contains the information needed
    *                      for inline evaluation
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runExpression(
    ir: IR.Expression,
    inlineContext: InlineContext
  ): IR.Expression = ir

  /** @inheritdoc */
  override def updateMetadataInDuplicate[T <: IR](
    @unused sourceIr: T,
    copyOfIr: T
  ): T = copyOfIr
}

File: engine/runtime/src/main/java/org/enso/compiler/pass/analyse/ImportApiAnalysis.java
package org.enso.compiler.pass.analyse;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.UUID;
import org.enso.compiler.context.InlineContext;
import org.enso.compiler.context.ModuleContext;
import org.enso.compiler.core.IR;
import org.enso.compiler.core.IR$Error$ImportExport;
import org.enso.compiler.core.IR$Error$ImportExport$SymbolDoesNotExist;
import org.enso.compiler.core.IR$Module$Scope$Import$Module;
import org.enso.compiler.data.BindingsMap;
import org.enso.compiler.data.BindingsMap$ModuleReference$Concrete;
import org.enso.compiler.pass.IRPass;
import org.enso.compiler.pass.desugar.ComplexType$;
import org.enso.compiler.pass.desugar.FunctionBinding$;
import org.enso.compiler.pass.desugar.GenerateMethodBodies$;
import org.enso.compiler.pass.resolve.MethodDefinitions$;
import org.enso.compiler.pass.resolve.Patterns$;
import org.enso.interpreter.runtime.EnsoContext;
import org.enso.interpreter.runtime.Module;
import org.enso.interpreter.util.ScalaConversions;
import org.enso.pkg.QualifiedName;
import scala.collection.immutable.Seq;

/** Verifies that all imported symbols are also accessible via {@code Main.enso}.
 * Checks if a package has {@code Main.enso} file and if so it replaces all
 * imports that bypass that file with an error.
 */
public class ImportApiAnalysis implements IRPass {
  public static final ImportApiAnalysis MODULE$ = new ImportApiAnalysis();
  private UUID uuid;

  private ImportApiAnalysis() {
  }

  @Override
  public void org$enso$compiler$pass$IRPass$_setter_$key_$eq(UUID v) {
    this.uuid = v;
  }

  @Override
  public UUID key() {
    return uuid;
  }

  @Override
  public Seq<IRPass> precursorPasses() {
    return cons(ComplexType$.MODULE$, cons(FunctionBinding$.MODULE$, cons(GenerateMethodBodies$.MODULE$, nil())));
  }

  @Override
  public Seq<IRPass> invalidatedPasses() {
    return cons(MethodDefinitions$.MODULE$, cons(Patterns$.MODULE$, nil()));
  }

  @Override
  public IR.Module runModule(IR.Module ir, ModuleContext moduleContext) {
    var map = (BindingsMap) ir.passData().get(BindingAnalysis$.MODULE$).get();
    var forbiddenImports = new HashMap<String, String>();

    for (var imp : ScalaConversions.asJava(map.resolvedImports())) {
      var mod = switch (imp.target()) {
        case BindingsMap.ResolvedModule rm ->
          switch (rm.module()) {
            case BindingsMap$ModuleReference$Concrete c ->
              c.module();
            default ->
              null;
          };
        default ->
          null;
      };

      if (mod != null) {
        var pkg = mod.getPackage();
        if (moduleContext.module().getPackage() != pkg) {
          if (pkg.mainFile() != null) {
            if (pkg.findModule("Main") instanceof Module mainModule) {
              // if the package has Main file, then imports has to go thru the main mod
              boolean forbidden;
              if (mod == mainModule) {
                forbidden = false;
              } else {
                forbidden = !findLogicalImport(mod, mainModule);
              }
              if (forbidden) {
                forbiddenImports.put(
                        mod.getName().toString(), mainModule.getName().toString()
                );
              }
            }
          }
        }
      }
    }

    var replace = ir.imports().map((imp) -> switch (imp) {
      case IR$Module$Scope$Import$Module mod -> {
        var name = mod.name().name();
        var pkg = forbiddenImports.get(name);
        if (pkg != null) {
          var segments = name.split("\\.");
          var last = segments[segments.length - 1];
          yield new IR$Error$ImportExport(
            imp,
            new IR$Error$ImportExport$SymbolDoesNotExist(last, pkg),
            imp.passData(), imp.diagnostics()
          );
        } else {
          yield mod;
        }
      }
      default ->
        imp;
    });
    return ir.copy(
            replace,
            ir.copy$default$2(),
            ir.copy$default$3(),
            ir.copy$default$4(),
            ir.copy$default$5(),
            ir.copy$default$6(),
            ir.copy$default$7()
    );
  }

  @Override
  public IR.Expression runExpression(IR.Expression ir, InlineContext inlineContext) {
    return ir;
  }

  @Override
  public <T extends IR> T updateMetadataInDuplicate(T sourceIr, T copyOfIr) {
    return copyOfIr;
  }

  /** @return {@code true} if the {@code mod} isn't allowed to be imported */
  private static boolean findLogicalImport(Module mod, Module rootModule) {
    var checked = new HashSet<>();
    var toProcess = new LinkedList<Module>();
    toProcess.add(rootModule);

    for (;;) {
      if (toProcess.isEmpty()) {
        break;
      }
      var current = toProcess.remove();
      if (!checked.add(current)) {
        continue;
      }
      var ir = current.getIr();
      if (ir == null) {
        var ctx = EnsoContext.get(null);
        ir = current.compileScope(ctx).getModule().getIr();
      }
      var meta = ir.passData();
      var map = (BindingsMap) meta.get(BindingAnalysis$.MODULE$).get();

      var itMainMap = map.exportedSymbols().iterator();
      while (itMainMap.hasNext()) {
        for (var module : ScalaConversions.asJava(itMainMap.next()._2())) {
          switch (module) {
            case BindingsMap.ResolvedModule rm -> {
              switch (rm.module()) {
                case BindingsMap$ModuleReference$Concrete allowed -> {
                  if (allowed.module() == mod) {
                    // found import
                    return true;
                  } else {
                    if (isParentModule(allowed.getName(), mod.getName())) {
                      toProcess.add(allowed.module());
                    }
                  }
                }
                default -> {
                }
              }
            }
            default -> {
            }
          }
        }
      }
    }
    // no logical import found
    return false;
  }

  private static boolean isParentModule(QualifiedName parent, QualifiedName mod) {
    var at = 0;
    var parentPath = parent.pathAsJava();
    var modPath = mod.pathAsJava();
    while (at < parentPath.size()) {
      if (at >= modPath.size()) {
        return false;
      }
      if (!modPath.get(at).equals(parentPath.get(at))) {
        return false;
      }
      at++;
    }
    return modPath.size() > at && parent.item().equals(modPath.get(at));
  }

  @SuppressWarnings("unchecked")
  private static <T> scala.collection.immutable.List<T> nil() {
    return (scala.collection.immutable.List<T>) scala.collection.immutable.Nil$.MODULE$;
  }
  private static <T> scala.collection.immutable.List<T> cons(T head, scala.collection.immutable.List<T> tail) {
    return scala.collection.immutable.$colon$colon$.MODULE$.apply(head, tail);
  }
}
